### PR TITLE
Reinstate tags-as-metadata adding option to merge tags. Fixes #699

### LINKED ISSUE #699: Consider reinstating tags-as-metadata option
See #696 and #698

### PR DESCRIPTION
**Pull request by lucasoares**
Hello.

See #699 and linked issues for more information.

This is my first PR into spring-cloud-consul. To be honest I didn't explore the project due to lack of time so I wanted to make the few changes as possible. Please let me know if there is anything to change.

I also didn't had time to configure checkstyle so I will wait for for CI results.

> Just saw I authored commits using another user account. My bad.

Thanks.

### CODE REVIEW

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryProperties.java [hunk 1]
```diff
@@ -205,6 +205,12 @@
 	 */
 	private int order = 0;
 
+	/**
+	 * Enable merge consul tag data with metadata to create service instance (defaults
+	 * to false).
+	 */
+	private boolean enableTagMerge = false;
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryProperties.java:212]: rename to `mergeTagsEnabled`

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulServiceInstance.java [hunk 2]
```diff
@@ -25,23 +26,29 @@
...
 import org.springframework.core.style.ToStringCreator;
+import org.springframework.util.StringUtils;
 
 import static org.springframework.cloud.consul.discovery.ConsulServerUtils.findHost;
 
 public class ConsulServiceInstance extends DefaultServiceInstance {
 
 	private HealthService healthService;
 
-	public ConsulServiceInstance(HealthService healthService, String serviceId) {
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulServiceInstance.java:35]: This is a public API so existing constructor signatures will need to remain with a default value.

**Review by spencergibb — CHANGES_REQUESTED** [review]
If you look at 2a667f48e0dcb665886b85796597ff947f181bba055c1216d573b68cb3764a8c, f67284561e873590ee7505776aa436501faff13a and 64852409228f1642fa65117e7c382730efbf37f8 there is some more to do to get back to parity (though merged rather than replaced).

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryProperties.java (hunk 1 above)
Comment (lucasoares) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulDiscoveryProperties.java:212]: Done.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulServiceInstance.java (hunk 2 above)
Comment (lucasoares) [reply spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/ConsulServiceInstance.java:35]: There is already a constructor with default merge value as `false`.

I just moved the default constructor to be the first in the file.

**Comment by lucasoares** [comment]
> If you look at 2a667f48e0dcb665886b85796597ff947f181bba055c1216d573b68cb3764a8c, [f672845](https://github.com/spring-cloud/spring-cloud-consul/commit/f67284561e873590ee7505776aa436501faff13a) and [6485240](https://github.com/spring-cloud/spring-cloud-consul/commit/64852409228f1642fa65117e7c382730efbf37f8) there is some more to do to get back to parity (though merged rather than replaced).

Looking into it. I solved other discussions but not committed yet.

I will remove draft status when ready.

**Comment by lucasoares** [comment]
> If you look at 2a667f48e0dcb665886b85796597ff947f181bba055c1216d573b68cb3764a8c, [f672845](https://github.com/spring-cloud/spring-cloud-consul/commit/f67284561e873590ee7505776aa436501faff13a) and [6485240](https://github.com/spring-cloud/spring-cloud-consul/commit/64852409228f1642fa65117e7c382730efbf37f8) there is some more to do to get back to parity (though merged rather than replaced).

Looking into it.

I will remove draft status when ready.

@spencergibb on [f672845](https://github.com/spring-cloud/spring-cloud-consul/commit/f67284561e873590ee7505776aa436501faff13a) we should set both tags and metadata if the merge is enabled? Because before we had a property for tags **or** metadata. Now we have a property for tags **and** metadata. What I should do?

**Comment by lucasoares** [comment]
> If you look at 2a667f48e0dcb665886b85796597ff947f181bba055c1216d573b68cb3764a8c, [f672845](https://github.com/spring-cloud/spring-cloud-consul/commit/f67284561e873590ee7505776aa436501faff13a) and [6485240](https://github.com/spring-cloud/spring-cloud-consul/commit/64852409228f1642fa65117e7c382730efbf37f8) there is some more to do to get back to parity (though merged rather than replaced).

Looking into it.

I will remove draft status when ready.

@spencergibb on [f672845](https://github.com/spring-cloud/spring-cloud-consul/commit/f67284561e873590ee7505776aa436501faff13a) we should set both tags and metadata if the merge is enabled? Because before we had a property for tags **or** metadata. Now we have a property for tags **and** metadata. What I should do?

**Comment by spencergibb** [comment]
Yes, I think it should be a merge.

**Comment by spencergibb** [comment]
@lucasoares any chance to update this?

**Comment by lucasoares** [comment]
> @lucasoares any chance to update this?

Hello @spencergibb.

Unfortunately, I had no time to check it. I will try to get back to it this weekend.

**Comment by lucasoares** [comment]
Closing in favor of https://github.com/spring-cloud/spring-cloud-consul/pull/775

**Pull request closed without merging**